hir/src/function.rs

```diff
@@ -62,6 +62,13 @@ pub enum CallConv {
     ///
     /// In all other respects, this calling convention is the same as `SystemV`
     Kernel,
+    /// A function with this calling convention is expected to be called using the `call`
```

Comment:
Note: Per the [Component Translation](https://github.com/0xPolygonMiden/compiler/discussions/360) discussion and our conversation today - it isn't necessary to represent the concept of context switching as an ABI unto itself, rather it is just one property of an ABI - arguably it is a property of a specific call site, not the callee itself, but for our purposes deriving it from ABI information works. I suggest that instead of `CrossCtx`, we specify three ABIs for use by our Wasm frontend: `Wasm`, `CanonLift`, and `CanonLower` (corresponding to core Wasm ABI and the respective primitives specified in the [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#canon-lift)).

To elaborate on this:

* `Wasm` we already have of course, but it doesn't _really_ mean anything, and perhaps could be named something more generic. Essentially, this calling convention mostly indicates that: 
  a.) It came from a core Wasm module
  b.) We don't know anything about the "real" types of the arguments/results, i.e. it is mostly opaque from a type system perspective
  c.) The actual calling convention details of the function are encoded in its signature - by which I mean that if a specific convention was expected by the original source language, it has been compiled to Wasm such that those details are already baked in and represented by the core Wasm types that remain. Where it is unclear what the semantics are for a given type, we should consult the core Wasm spec for the `call` or `call_indirect` instructions.
* `CanonLift` and `CanonLower` correspond to their respective primitives as described in the [Canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#canon-lift) spec.
* Functions with the `CanonLift` ABI represent the actual component boundary, and calling/returning from them involves the actual context switch in the Miden VM. We'll revisit that part in a moment. This ABI is also the only one whose implementation details are _not_ fully specified by the Component Model/Canonical ABI. While the semantics of `canon lift` are described, the implementation is host-defined, and therefore is ours to dictate. This works out great for us, because this also happens to be where the idiosyncratic details of Miden's context switching must be dealt with, so this gives us flexibility.
* `CanonLower` is the only ABI other than `Wasm` that is callable from a `Wasm` ABI function. The actual details of this ABI are identical to `Wasm` from the perspective of a caller, however the use of a unique ABI for these functions allows us to identify them as synthetic, and to ensure that calls to a `CanonLift` function only occur via its `CanonLower` counterpart. A call to a `CanonLower` function can be thought of as a call to another `Wasm` ABI function, but with a layer of lifting/lowering inbetween, so what we end up with is actually a chain of calls composed of all three ABIs, i.e. `Wasm` (caller) -> `CanonLower` (wraps Canonical ABI function) -> `CanonLift` (wraps core function) -> `Wasm` (callee). A `CanonLower` function is _always_ paired with a `CanonLift` function, and is only ever called from a `Wasm` function. Conversely, `CanonLift` functions are never called directly from a `Wasm` function, only from a `CanonLower` function.
* Neither `Wasm` nor `CanonLower` require us to emit any special calling convention-related code, the details of the Canonical ABI itself are materialized in the implementation of the `CanonLower` and `CanonLift` functions.
* Both `Wasm` and `CanonLower` functions have type signatures consisting solely of core Wasm types.
* `CanonLift` functions have type signatures consisting solely of Component Model types. For the most part, we don't need to concern ourselves with these types, _except_ in regard to how to pass them across the context switch (i.e. as arguments in the call from `CanonLower` -> `CanonLift`, and as results when returning from `CanonLift`). The lifting/lowering of these types is described [here](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#lifting-and-lowering-values), but the exact way in which we pass them across the `CanonLift` boundary is up to us.

With all that in mind, the only thing we need to determine if a function invocation should be lowered as `exec` or `call`, is to check the ABI of the callee. If it is `CanonLift`, then we use `call`, all other function invocations will use `exec`. As mentioned above, we also will validate all call sites to ensure that calls to `CanonLower` and `CanonLift` functions do not originate from invalid callers (e.g. direct from `Wasm` -> `CanonLift`). I should note that `CanonLower` and `CanonLift`, in terms of ABIs, _also_ encompass all of the other semantics as described in the Canonical ABI spec, so the fact that `CanonLift` corresponds to a context boundary is only one part of its contract.

Reply:
> The `canon lift` function description has CM type signatures, but the generated CanonLift function during the account code compilation has a "lowered" Miden cross-context ABI signature. 

It only has a single signature, the "high-level" type signature, there is no low-level signature for the `CanonLift` function at all. This is because it is called _only_ from a corresponding `CanonLower` function. The mechanics of the calling convention (when calling from `CanonLower` -> `CanonLift`) for those high-level types is up to us, i.e. how do we pass/return e.g. a `list`? However we see fit, because the `CanonLift` calling convention is host-defined. This calling convention is also where the context switch occurs, and the mechanics of that are also ours to define.

> For example, for the CM signature (list<felt>) -> list<felt> we will generate a synthetic function that uses advice provider, pass hashes and will have the "lowered" Miden CCABI (word) -> word signature.

Ah, I see where the confusion is coming from. You are assuming that we need to represent the calling convention implementation details in the function signature, resulting in needing two signatures (one to preserve the original information, one to encode the implementation details). This is not necessary though. The implementation details of the calling convention are _always_ derived from the original signature, so it is not necessary for us to encode them as a second signature, we just derive the latter as needed (essentially only during codegen). To be clear here, there is no "Miden ABI" or "Miden calling convention" per se - there _are_ constraints on how we can execute procedure calls in the Miden VM, and so any calling convention must be lowered to MASM in such a way that it works within those constraints. The "lowering" of calling conventions into Miden Assembly is responsible for emitting code that preserves the original calling convention semantics, but adheres to Miden's constraints at the VM level.

There are two places where we might want to do the actual work of preparing a function so that it is "Miden-compatible", by which I mean it can be lowered to Miden Assembly directly following the normal type layout rules, without violating any of the procedure invocation constraints:

1. During code generation, which is where the vast majority of these details will be handled, in particular all of the code related to laying out types in memory, interacting with the advice provider, operand stack management, etc.
2. As an IR transformation prior to codegen, so that we can ensure that code motion affecting function arguments is accounted for when we run the spills transformation. For example, if we have a `CanonLift` function whose signature is `list<felt> -> list<felt>`, we must rewrite the function (and any call sites) as if the signature was `digest -> digest`, as heap-allocated types cannot be passed across a context switch, so we instead store the lists in the advice provider, and pass the key in the advice map from which the actual list can be retrieved. The prologue of the function will then use that digest to extract the list into memory, and replace any uses of the original function argument with the list value once it has been read out of the advice provider. It should be noted that this transformation is implied/derived from the original signature, so we do not need to have the actual function definition to rewrite call sites, nor do we need to worry about call sites when rewriting the function.

The important thing to note about 2 is that while the actual IR is rewritten, the type signature is left unmodified! This is because the rewrite performed here doesn't change the semantics of the function at all, it is purely an artifact of representing those semantics in Miden Assembly.

I should also note that we do, in a sense, store both the original signature, and the "effective" signature here, because in HIR2 a `Function` has a `signature` attribute (a `Signature` value, corresponding to the original signature, and also providing things like ABI details), but we can also directly interrogate the types of its arguments and results as represented in the IR at any given point in time. However in almost all cases, the two will be the same. The only time we'd allow them to diverge is during codegen, in order to facilitate the rewrite I mentioned above.

Reply:
> > The `canon lift` function description has CM type signatures, but the generated CanonLift function during the account code compilation has a "lowered" Miden cross-context ABI signature.
> 
> It only has a single signature, the "high-level" type signature, there is no low-level signature for the `CanonLift` function at all. This is because it is called _only_ from a corresponding `CanonLower` function. The mechanics of the calling convention (when calling from `CanonLower` -> `CanonLift`) for those high-level types is up to us, i.e. how do we pass/return e.g. a `list`? However we see fit, because the `CanonLift` calling convention is host-defined. This calling convention is also where the context switch occurs, and the mechanics of that are also ours to define.
> 
> > For example, for the CM signature (list) -> list we will generate a synthetic function that uses advice provider, pass hashes and will have the "lowered" Miden CCABI (word) -> word signature.
> 
> Ah, I see where the confusion is coming from. You are assuming that we need to represent the calling convention implementation details in the function signature, resulting in needing two signatures (one to preserve the original information, one to encode the implementation details). This is not necessary though. The implementation details of the calling convention are _always_ derived from the original signature, so it is not necessary for us to encode them as a second signature, we just derive the latter as needed (essentially only during codegen). To be clear here, there is no "Miden ABI" or "Miden calling convention" per se - there _are_ constraints on how we can execute procedure calls in the Miden VM, and so any calling convention must be lowered to MASM in such a way that it works within those constraints. The "lowering" of calling conventions into Miden Assembly is responsible for emitting code that preserves the original calling convention semantics, but adheres to Miden's constraints at the VM level.

This helps and clarifies a lot. Thank you for the detailed explanation! It started to click for me. The high-level signature is surely enough to derive the low-level "implementation detail" one.

> 
> There are two places where we might want to do the actual work of preparing a function so that it is "Miden-compatible", by which I mean it can be lowered to Miden Assembly directly following the normal type layout rules, without violating any of the procedure invocation constraints:
> 
> 1. During code generation, which is where the vast majority of these details will be handled, in particular all of the code related to laying out types in memory, interacting with the advice provider, operand stack management, etc.
> 2. As an IR transformation prior to codegen, so that we can ensure that code motion affecting function arguments is accounted for when we run the spills transformation. For example, if we have a `CanonLift` function whose signature is `list<felt> -> list<felt>`, we must rewrite the function (and any call sites) as if the signature was `digest -> digest`, as heap-allocated types cannot be passed across a context switch, so we instead store the lists in the advice provider, and pass the key in the advice map from which the actual list can be retrieved. The prologue of the function will then use that digest to extract the list into memory, and replace any uses of the original function argument with the list value once it has been read out of the advice provider. It should be noted that this transformation is implied/derived from the original signature, so we do not need to have the actual function definition to rewrite call sites, nor do we need to worry about call sites when rewriting the function.
> 
> The important thing to note about 2 is that while the actual IR is rewritten, the type signature is left unmodified! This is because the rewrite performed here doesn't change the semantics of the function at all, it is purely an artifact of representing those semantics in Miden Assembly.

This part confuses me. From the previous comment, I understood that `CanonLift` and `CanonLower` synthetic functions are generated in the frontend, and now it appears that they are "implemented" in the IR transformation phase or the codegen. This leads me to think that the frontend generates empty synthetic functions which later "implemented" with the "Miden-compatible" lowering you described above either in the IR transformation phase or the codegen. Am I on the right track?

Reply:
There are a couple different things here being accidentally conflated:

* A `CanonLower` function _declaration_ presents as just another `Wasm` function, to `Wasm` callers, from a semantics point of view. However, what it really doing is adapting a function with the `CanonLift` calling convention to the `Wasm` convention. An alternate way of thinking about it, is that we could allow `Wasm` functions to directly call `CanonLift` functions, but then we would need to emit the equivalent of the `CanonLower` implementation at every call site. This would generate a lot more code for no benefit.
* The _implementation_ of a `CanonLower` function emitted by the frontend must handle translating core Wasm types into their Canonical ABI representation. This could be represented with something like `cast`, but may need to be more explicit (depending on the type, I'm not sure). It then calls the corresponding `CanonLift` function with arguments in the Canonical ABI representation. Note that, at this point, no low-level calling convention details are being represented - only the type representation from core Wasm to Canonical ABI.
* The `CanonLift` _declaration_ is only callable from a `CanonLower` function, which can be thought of as a Canonical ABI caller (since internally it has converted types to the Canonical ABI). Much like how `CanonLower` presents as a `Wasm` function from the perspective of a `Wasm` caller `CanonLift` presents as a Canonical ABI function to a Canonical ABI caller, by internally adapting the underlying `Wasm` calling convention function to the Canonical ABI.
* The _implementation_ of a `CanonLift` function emitted by the frontend must handle translating Canonical ABI types into their core Wasm representation. Again, this could be represented with `cast`, but likely requires more explicit lowering code for certain types. Once complete, it then calls the underlying `Wasm` function. Again, no low-level calling convention details are being represented here - just the type representation change.

So in short:

* The frontend synthesizes the `CanonLift`/`CanonLower` functions with code that effects the type representation change
* The backend handles the low-level calling convention details (e.g. how Canonical ABI types are passed for the `CanonLift` convention, how to handle context switches, operand stack management, laying out types in memory, etc.)

Reply:
I should also note that, we can't represent the type representation change purely using `cast` (except where the cast is obviously applicable, e.g. between integral types and such), because the IR type system doesn't distinguish between WebAssembly and Canonical ABI types, and shouldn't. Casts are essentially representation-agnostic, expressing only a conversion between IR types, and the default Miden ABI layout of those types may not match what is specified by the Canonical ABI. 

So by emitting (in the frontend) code to handle the representation change, we ensure that the code correctly lays things out according to the Canonical ABI rules, and then during codegen, we will know that the IR types corresponding to the function arguments/results are in Canonical ABI representation (as it is implied by the calling convention), and we will emit the low-level details of transferring arguments/results in that representation from caller to callee. In other words, the combination of IR type and calling convention imply a specific data layout when used as a function argument or result.

The Wasm aspect of all this is a bit weird. Basically, the Component Model is able to avoid explicitly representing the type representation change in the form of Wasm instructions because the host runtime speaks Wasm and can handle it automatically (and therefore more efficiently). Typically though, adapting types from one type system to another would require the frontend emit code to do that work somewhere, and then the low-level details are left up to the backend. We're in kind of an awkward position, as we must act as both frontend and host runtime: First, we materialize functions that perform the work that would normally have been emitted by a frontend on more traditional targets, to effect the type representation change. Second, we act as the host, by enforcing the shared-nothing boundary via context switching, and managing the movement of data across that boundary. Where it gets confusing is that, in the functions we synthesize to do this, we have more traditional calling convention details in the mix as well. So it gets a bit fuzzy here, in terms of being precise about what parts of the apparent calling convention are handled by the frontend, and what is handled by the backend. Hopefully my commentary here has made that distinction clearer, but I certainly can't blame you if it is still confusing.

Reply:
Now, I feel that it finally clicked for me! Thank you for the detailed explanation!